eras/babbage/impl/src/Cardano/Ledger/Babbage/TxBody.hs

```diff
@@ -0,0 +1,805 @@
+
+outputs' (TxBodyConstr (Memo raw _)) = _outputs raw
+
+certs' (TxBodyConstr (Memo raw _)) = _certs raw
+
+wdrls' (TxBodyConstr (Memo raw _)) = _wdrls raw
+
+txfee' (TxBodyConstr (Memo raw _)) = _txfee raw
+
+vldt' (TxBodyConstr (Memo raw _)) = _vldt raw
+
+update' (TxBodyConstr (Memo raw _)) = _update raw
+
+reqSignerHashes' (TxBodyConstr (Memo raw _)) = _reqSignerHashes raw
+
+adHash' (TxBodyConstr (Memo raw _)) = _adHash raw
+
+mint' (TxBodyConstr (Memo raw _)) = _mint raw
+
+scriptIntegrityHash' (TxBodyConstr (Memo raw _)) = _scriptIntegrityHash raw
+
+txnetworkid' (TxBodyConstr (Memo raw _)) = _txnetworkid raw
+
+--------------------------------------------------------------------------------
+-- Serialisation
+--------------------------------------------------------------------------------
+
+instance
+  ( Era era,
+    Compactible (Core.Value era)
+  ) =>
+  ToCBOR (TxOut era)
+  where
+  toCBOR (TxOutCompact addr cv) =
+    encodeListLen 2
+      <> toCBOR addr
+      <> toCBOR cv
+  toCBOR (TxOutCompactDH addr cv dh) =
+    encodeListLen 3
+      <> toCBOR addr
+      <> toCBOR cv
+      <> toCBOR dh
+  toCBOR x =
+    let (addr, cv, dh) = viewCompactTxOut x
+     in encodeListLen 3
+          <> toCBOR addr
+          <> toCBOR cv
+          <> toCBOR dh
+
+instance
+  ( Era era,
+    DecodeNonNegative (Core.Value era),
+    Show (Core.Value era),
+    Compactible (Core.Value era)
+  ) =>
+  FromCBOR (TxOut era)
+  where
+  fromCBOR = do
+    lenOrIndef <- decodeListLenOrIndef
+    case lenOrIndef of
+      Nothing -> do
+        a <- fromCBOR
+        cv <- decodeNonNegative
+        decodeBreakOr >>= \case
+          True -> pure $ TxOutCompact a cv
+          False -> do
+            dh <- fromCBOR
+            decodeBreakOr >>= \case
+              True -> pure $ TxOutCompactDH a cv dh
+              False -> cborError $ DecoderErrorCustom "txout" "Excess terms in txout"
+      Just 2 ->
+        TxOutCompact
+          <$> fromCBOR
+          <*> decodeNonNegative
+      Just 3 ->
+        TxOutCompactDH
+          <$> fromCBOR
+          <*> decodeNonNegative
+          <*> fromCBOR
+      Just _ -> cborError $ DecoderErrorCustom "txout" "wrong number of terms in txout"
+
+encodeTxBodyRaw ::
+  ( Era era,
+    ToCBOR (PParamsDelta era)
+  ) =>
+  TxBodyRaw era ->
+  Encode ('Closed 'Sparse) (TxBodyRaw era)
+encodeTxBodyRaw
+  TxBodyRaw
+    { _inputs,
+      _collateral,
+      _collateralReturn,
+      _outputs,
+      _certs,
+      _wdrls,
+      _txfee,
+      _vldt = ValidityInterval bot top,
+      _update,
+      _reqSignerHashes,
+      _mint,
+      _scriptIntegrityHash,
+      _adHash,
+      _txnetworkid
+    } =
+    Keyed
+      ( \i ifee ca o f t c w u b rsh mi sh ah ni ->
+          TxBodyRaw i ifee ca o c w f (ValidityInterval b t) u rsh mi sh ah ni
+      )
+      !> Key 0 (E encodeFoldable _inputs)
+      !> Key 13 (E encodeFoldable _collateral)
+      !> Key 13 (To _collateralReturn)
```

Comment:
Correct me if I am wrong, but I assumed that key should be unique and reflect the position of the field during serlization/deserialization.